editor/xml_feature.cpp

```diff
@@ -409,6 +409,135 @@ string XMLFeature::GetUploadError() const { return GetRootNode().attribute(kUplo
 
 void XMLFeature::SetUploadError(string const & error) { SetAttribute(kUploadError, error); }
 
+osm::EditJournal XMLFeature::GetEditJournal() const
+{
+  // debug print
+  std::ostringstream ost;
+  Save(ost);
+  LOG(LDEBUG, ("Read in XMLFeature:\n", ost.str()));
+
+  auto readEditJournalList = [] (pugi::xml_node & xmlNode, osm::EditJournal & journal, bool isHistory)
+  {
+    for (auto xmlEntry = xmlNode.child("entry"); xmlEntry; xmlEntry = xmlEntry.next_sibling("entry"))
+    {
+      osm::JournalEntry entry;
+
+      std::string strType = xmlEntry.attribute("type").value();
+      std::optional<osm::JournalEntryType> type = osm::EditJournal::TypeFromString(strType);
+      ASSERT(type.has_value(), ("Invalid JournalEntryType ", strType));
+      entry.journalEntryType = *type;
+
+      entry.timestamp = base::StringToTimestamp(xmlEntry.attribute("timestamp").value());
+
+      auto xmlData = xmlEntry.child("data");
+
+      switch (entry.journalEntryType) {
+        case osm::JournalEntryType::TagModification:
+        {
+          osm::TagModData tagModData;
+          tagModData.key = xmlData.attribute("key").value();
+          tagModData.old_value = xmlData.attribute("old_value").value();
+          tagModData.new_value = xmlData.attribute("new_value").value();
+          entry.data = tagModData;
+          break;
+        }
+        case osm::JournalEntryType::ObjectCreated:
+        {
+          osm::ObjCreateData objCreateData;
+          objCreateData.type = classif().GetTypeByReadableObjectName(xmlData.attribute("type").value());
+          objCreateData.geomType = feature::TypeFromString(xmlData.attribute("geomType").value());
+          objCreateData.mercator = mercator::FromLatLon(GetLatLonFromNode(xmlData));
+          entry.data = objCreateData;
+          break;
+        }
+        case osm::JournalEntryType::LegacyObject:
+        {
+          osm::LegacyObjData legacyObjData;
+          legacyObjData.version = xmlData.attribute("version").value();
+          entry.data = legacyObjData;
+          break;
+        }
+      }
```

Comment:
What will happen if xml file is broken, say any XXX.value() does not exist? Should we have any generic error processing here? Like catch an exception and continue/skip the whole entry?

Reply:
pugixml shouldn't throw exceptions. It's very fault tolerant, so e.g. calling `.attribute("key").value()` on an empty xml node doesn't crash, but just returns "".

The current `GetEditJournal()` assumes that the xml is correct. Should I add checks and use MYTHROW in case read in values are invalid like it is done in other parts of the `xml_feature` class?

Reply:
I added checks to make sure only correct values are read from the xml file. (exceptions are thrown when the xml contains an invalid value)